velox/dwio/common/tests/TestParallelBufferedInput.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/executors/IOThreadPoolExecutor.h>
+#include <gtest/gtest.h>
+#include "velox/dwio/common/ParallelBufferedInput.h"
+#include "velox/exec/tests/utils/TempFilePath.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::dwio::common;
+
+TEST(TestParallelBufferedInput, parallelLoad) {
+  const int32_t kOneKB = 1 * 1024;
+  const int32_t kOneMB = 1 * 1024 * 1024;
+  const int32_t threads = 4;
+
+  auto ioStats = std::make_shared<IoStatistics>();
+  auto executor = std::make_unique<folly::IOThreadPoolExecutor>(threads);
+  int32_t loadQuantum = kOneKB;
+
+  std::string fileData;
+  {
+    InMemoryWriteFile writeFile(&fileData);
+    writeFile.append(std::string(kOneMB, 'a'));
+    writeFile.append(std::string(2 * kOneMB, 'b'));
+    writeFile.append(std::string(2 * kOneMB, 'c'));
+    writeFile.append(std::string(kOneMB, 'd'));
+  }
+  auto readFile = std::make_shared<InMemoryReadFile>(fileData);
+  auto pool = memory::getDefaultMemoryPool();
+  ParallelBufferedInput input(
+      readFile,
+      *pool,
+      MetricsLog::voidLog(),
+      ioStats,
+      executor.get(),
+      loadQuantum);
+
+  // Avoid coalescing.
+  std::vector<Region> regions = {
+      {0, kOneKB},
+      {2 * kOneMB, 10},
+      {4 * kOneMB, readFile->size() - (4 * kOneMB)}};
+  std::vector<std::unique_ptr<SeekableInputStream>> streams;
+  for (auto& region : regions) {
+    streams.push_back(input.enqueue(region));
+  }
+
+  input.load(LogType::TEST);
+
+  for (size_t i = 0; i < regions.size(); i++) {
+    const auto& region = regions[i];
+    const auto& stream = streams[i];
+
+    auto buffer =
+        std::make_shared<dwio::common::DataBuffer<char>>(*pool, region.length);
+    stream->readFully(buffer->data(), region.length);
+
+    // Verify data correctness.
+    auto expect = readFile->pread(region.offset, region.length);
+    ASSERT_EQ(buffer->size(), expect.size());
+    ASSERT_EQ(buffer->data(), expect);
+  }
+
+  int32_t totalLength = 0;
+  for (const auto& region : regions) {
+    totalLength += region.length;
+  }
+  ASSERT_EQ(totalLength, ioStats->rawBytesRead());
+
+  ASSERT_EQ(0, ioStats->rawOverreadBytes());
+}
+
+class TestReadFile : public ReadFile {
```

Comment:
Can you derive this from `InMemoryReadFile` to simplify and avoid defining other methods?

Reply:
Let's test with coalescing in this test since the previous one avoided it.

Reply:
Add `mergeRegions` to test coalescing.